jsprit-core/src/main/java/com/graphhopper/jsprit/core/algorithm/recreate/Inserter.java

```diff
@@ -17,117 +17,116 @@
-        }
+    class JobExceptionHandler extends JobInsertionHandler {
 
         @Override
         public void handleJobInsertion(Job job, InsertionData iData, VehicleRoute route) {
-            if (job instanceof Service) {
-                route.setVehicleAndDepartureTime(iData.getSelectedVehicle(), iData.getVehicleDepartureTime());
-                if (!iData.getSelectedVehicle().isReturnToDepot()) {
-                    if (iData.getDeliveryInsertionIndex() >= route.getTourActivities().getActivities().size()) {
-                        setEndLocation(route, (Service) job);
-                    }
-                }
-                TourActivity activity = vehicleRoutingProblem.copyAndGetActivities(job).get(0);
-                route.getTourActivities().addActivity(iData.getDeliveryInsertionIndex(), activity);
-            } else delegator.handleJobInsertion(job, iData, route);
-        }
-
-        private void setEndLocation(VehicleRoute route, Service service) {
-            route.getEnd().setLocation(service.getLocation());
-        }
-
-        public void setNextHandler(JobInsertionHandler jobInsertionHandler) {
-            this.delegator = jobInsertionHandler;
+            throw new IllegalStateException("job insertion is not supported. Do not know job type.");
         }
 
     }
 
-    class ShipmentInsertionHandler implements JobInsertionHandler {
-
-        private final VehicleRoutingProblem vehicleRoutingProblem;
 
-        private TourShipmentActivityFactory activityFactory = new DefaultShipmentActivityFactory();
+    private class UnifiedInsertionHandler extends JobInsertionHandler {
 
-        private JobInsertionHandler delegator = new JobExceptionHandler();
-
-        public ShipmentInsertionHandler(VehicleRoutingProblem vehicleRoutingProblem) {
-            this.vehicleRoutingProblem = vehicleRoutingProblem;
+        public UnifiedInsertionHandler() {
         }
 
         @Override
         public void handleJobInsertion(Job job, InsertionData iData, VehicleRoute route) {
-            if (job instanceof Shipment) {
-                List<AbstractActivity> acts = vehicleRoutingProblem.copyAndGetActivities(job);
-                TourActivity pickupShipment = acts.get(0);
-                TourActivity deliverShipment = acts.get(1);
-                route.setVehicleAndDepartureTime(iData.getSelectedVehicle(), iData.getVehicleDepartureTime());
-                if (!iData.getSelectedVehicle().isReturnToDepot()) {
-                    if (iData.getDeliveryInsertionIndex() >= route.getActivities().size()) {
-                        setEndLocation(route, (Shipment) job);
-                    }
+            route.setVehicleAndDepartureTime(iData.getSelectedVehicle(), iData.getVehicleDepartureTime());
+            if (!iData.getSelectedVehicle().isReturnToDepot()) {
+                if (iData.getDeliveryInsertionIndex() >= route.getActivities().size()) {
+                    route.getEnd().setLocation(job.getEndLocation());
```

Comment:
But I aggree with you, we should refactor this. Here we might wait until we know how we actually handle the insertion (insertion cost calculation) of multiple dependencies.